src/virtio_console.js

```diff
@@ -0,0 +1,294 @@
+
+"use strict";
+
+var VIRTIO_CONSOLE_DEVICE_READY     = 0;
+var VIRTIO_CONSOLE_DEVICE_ADD       = 1;
+var VIRTIO_CONSOLE_DEVICE_REMOVE    = 2;
+var VIRTIO_CONSOLE_PORT_READY       = 3;
+var VIRTIO_CONSOLE_CONSOLE_PORT     = 4;
+var VIRTIO_CONSOLE_RESIZE           = 5;
+var VIRTIO_CONSOLE_PORT_OPEN        = 6;
+var VIRTIO_CONSOLE_PORT_NAME        = 7;
+
+var VIRTIO_CONSOLE_F_SIZE           = 0;
+var VIRTIO_CONSOLE_F_MULTIPORT      = 1;
+var VIRTIO_CONSOLE_F_EMERG_WRITE    = 2;
+
+/**
+ * @constructor
+ *
+ * @param {CPU} cpu
+ */
+function VirtioConsole(cpu, bus) {
+    /** @const @type {BusConnector} */
+    this.bus = bus;
+    this.rows = 25;
+    this.cols = 80;
+    this.ports = 4;
+
+    var queues = [
+        {
+            size_supported: 16,
+            notify_offset: 0,
+        },
+        {
+            size_supported: 16,
+            notify_offset: 1,
+        },
+        {
+            size_supported: 16,
+            notify_offset: 2,
+        },
+        {
+            size_supported: 16,
+            notify_offset: 3,
+        },
+    ];
+
+    for (let i = 1; i < this.ports; ++i)
+    {
+        queues.push({size_supported: 16, notify_offset: 0});
+        queues.push({size_supported: 8, notify_offset: 1});
+    }
+
+    /** @type {VirtIO} */
+    this.virtio = new VirtIO(cpu,
+    {
+        name: "virtio-console",
+        pci_id: 0x07 << 3,
+        device_id: 0x1043,
+        subsystem_device_id: 3,
+        common:
+        {
+            initial_port: 0xB800,
```

Comment:
One thing I was unsure about was these values.  I just picked some based on the ones used for p9 and they seemed to work.

Reply:
Yes, this will likely create conflicts with the virtio-fs device. Could you change it?

Reply:
I had added 0x1000 to them all, I just dont know what that means.  I tested it with both console and p9 up and both seem to work together.